src/api/cpp/seq.cpp

```diff
@@ -71,7 +71,10 @@ seq::seq(double begin, double end, double step): m_gfor(false)
     }
     if (end >= 0 && begin >= 0 && signbit(end-begin) != signbit(step))
         AF_THROW_ERR("Sequence is invalid", AF_ERR_ARG);
-        //AF_THROW("step must match direction of sequence");
+    if (step>0 && !(begin<end))
```

Comment:
This should probably say:
```c++
    if (step > 0 && begin > end)
```
Because for step 1, it is correct when `begin == end`. Still though, this will fail for e.g., `seq(2, af::end)` aka `seq(2, -1, 1)`.

Whether the sequence is a valid index depends on the shape of the array it is indexing. In other words, we have no idea whether index 15 is less than index `af::end - 10` aka `-11`. I believe only the signs can be checked during the construction of the sequence. E.g., something like:
```c++
    if (signbit(end) == signbit(begin) && signbit(end-begin) != signbit(step))
        AF_THROW_ERR("Sequence is invalid", AF_ERR_ARG);
    if (end < 0 && begin > 0 && step < 0)
        AF_THROW_ERR("Sequence is invalid", AF_ERR_ARG);
    if (end > 0 && begin < 0 && step > 0)
        AF_THROW_ERR("Sequence is invalid", AF_ERR_ARG);
```